source/extensions/filters/http/composite/filter.h

```diff
@@ -0,0 +1,121 @@
+#include "extensions/filters/http/composite/factory_wrapper.h"
+
+#include "absl/types/variant.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Composite {
+
+struct FactoryCallbacksWrapper;
+
+#define ALL_COMPOSITE_FILTER_STATS(COUNTER)                                                        \
+  COUNTER(filter_delegation_error)                                                                 \
+  COUNTER(filter_delegation_success)
+
+struct FilterStats {
+  ALL_COMPOSITE_FILTER_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+class Filter : public Http::StreamFilter, Logger::Loggable<Logger::Id::filter> {
+public:
+  explicit Filter(FilterStats& stats) : decoded_headers_(false), stats_(stats) {}
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap& trailers) override;
+  Http::FilterMetadataStatus decodeMetadata(Http::MetadataMap& metadata_map) override;
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override {
+    decoder_callbacks_ = &callbacks;
+  }
+  void decodeComplete() override;
+
+  // Http::StreamEncoderFilter
+  Http::FilterHeadersStatus encode100ContinueHeaders(Http::ResponseHeaderMap& headers) override;
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus encodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap& metadata_map) override;
+  void setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callbacks) override {
+    encoder_callbacks_ = &callbacks;
+  }
+  void encodeComplete() override;
+
+  // Http::StreamFilterBase
+  void onDestroy() override {
+    if (delegated_filter_) {
+      // We need to explicitly specify which base class to the conversion via due
+      // to the diamond inheritance between StreamFilter and StreamFilterBase.
+      static_cast<Http::StreamDecoderFilter&>(*delegated_filter_).onDestroy();
+    }
+  }
+
+  void onMatchCallback(const Matcher::Action& action) override;
+
+private:
+  friend FactoryCallbacksWrapper;
+
+  // Use these to track whether we are allowed to insert a specific kind of filter. These mainly
+  // serve to surface an easier to understand error, as attempting to insert a filter at a later
+  // time will result in various FM assertions firing.
+  // TODO(snowp): Instead of validating this via ASSERTs, we should be able to validate that the
+  // match tree is only going to fire when we can actually inject a filter.
```

Comment:
This is partially fixed, right? I was actually going to ask for ASSERTs also, so IMO it's not terrible to just keep this but should the comment be updated?